### PR TITLE
Fix not closing after data received

### PR DESCRIPTION
**Pull request by Nek-**
In many cases, we receive data and then the stream end. The `ReadableResourceStream` was not closing the stream on end as it's expected to be done.

This commit adds a test that exposes the problem and suggests a fix in the related class.

### CODE REVIEW

File: tests/ReadableResourceStreamTest.php [hunk 1]
```diff
@@ -203,7 +203,7 @@ public function testEmptyStreamShouldNotEmitData()
         $conn = new ReadableResourceStream($stream, $loop);
         $conn->on('data', $this->expectCallableNever());
 
-        $conn->handleData($stream);
+        $conn->handleData();
```
Comment (Nek-) [inline tests/ReadableResourceStreamTest.php]: Just fixing. This parameter does not exist and test were probably not updated after an update...

**Comment by clue** [comment]
@Nek- Thanks for reporting, but I'm not sure I can currently reproduce the problem you're seeing.

It's expected behavior that the stream emits a final `data` event with some data and then only detects the `end` in the next loop tick. Is this the problem you're seeing?

**Comment by Nek-** [comment]
@clue checkout the test case I suggested. It's failing in this case.

**Comment by clue** [comment]
@Nek- That's right. It should report an `end` event just fine the second time its `handleData()` method is called (that's the event-loop listener attached via `addReadStream()`). Can you confirm this? :+1:

**Comment by Nek-** [comment]
If I understand correctly what you are saying, I made a mistake in my test-case and it should be the following, right?
```diff
        $conn = new ReadableResourceStream($stream, $loop);
        $conn->on('data', function ($data) use (&$capturedData) {
            $capturedData = $data;
        });
        $hasEnded = false;
        $conn->on('end', function () use (&$hasEnded) {
            $hasEnded = true;
        });

        $conn->handleData();
+        $conn->handleData();
        $this->assertTrue($hasEnded);
        $this->assertSame("hello\n", $capturedData);
```

But this case is not reflecting the reality, if in the tests `handleData` is called by ourself (for test purpose) in the real world, it's called by the event loop, which will not call it one more time if there is no more data and the stream is finished.

You could basically replace my test case with the following:
```diff
        $conn = new ReadableResourceStream($stream, $loop);
        $conn->on('data', function ($data) use (&$capturedData) {
            $capturedData = $data;
        });
        $hasEnded = false;
        $conn->on('end', function () use (&$hasEnded) {
            $hasEnded = true;
        });

        while (!feof($stream)) {
            $conn->handleData();
        }

        $this->assertTrue($hasEnded);
        $this->assertSame("hello\n", $capturedData);
```

Without my changes, it will fail the same way. Because once the stream is ended the `handleData` is not called anymore.

**Comment by clue** [comment]
> If I understand correctly what you are saying, I made a mistake in my test-case and it should be the following, right?

Correct. The event-loop will invoke the readable listener when the file descriptor is in a "readable" state. This includes both the `data` event and an empty read which will emit the `end` event. I encourage you to try this in a simple example that uses an actual loop instance to reproduce this. This should also be covered by the existing test suite and we're not currently aware of any related issues.

I believe this has been answered, so I'm closing this for now. Please come back with more details if this problem persists and we can always reopen this :+1:

**Comment by Nek-** [comment]
@clue 

No, you miss-understand me. I was telling you that my PR is useful:
1. Take a look at [this example](https://gl.nekland.fr/Nek/test-sf-http-client/blob/master/test1.php), `end` is never called!
2. Take a further look at my code, it will call `handleData` only once, even with the loop!
3. That's exactly what does the event loop: only calling `handleData` when there is data to send, otherwise, it's not called (basically, on end of stream, handleData is not called)

This is a bug reporting as well as the related fix.

**Comment by clue** [comment]
@Nek- I can not currently reproduce the problem you're describing. I would be happy to take a look at this, can you provide a minimal test script to reproduce this locally?

This specific situation is tested in a couple of test cases, e.g. https://github.com/reactphp/stream/blob/1a849e70e1a421d0166ec18949846a2eec3c8bc5/tests/DuplexResourceStreamIntegrationTest.php#L330, can you verify this is really related to this stream component? The event-loop should also trigger when the stream is no longer readable, this is also covered in https://github.com/reactphp/event-loop/blob/88aaa77ad3ffce3ff22f1361d88c22ec85a50e23/tests/AbstractLoopTest.php#L65 and others.

**Comment by clue** [comment]
@Nek- I understand this situation can be frustrating – and I hope you understand I'm trying my best to help figure out what problem you're seeing and helping provide a solution. As of now, I can not see a problem in this project:

> The test I added

As discussed above, this test should call the event listener function (`handleData()`) twice. This is covered by a number of tests in this project and the underlying event loop, e.g. https://github.com/reactphp/event-loop/blob/88aaa77ad3ffce3ff22f1361d88c22ec85a50e23/tests/AbstractLoopTest.php#L65 linked above.

> https://gl.nekland.fr/Nek/test-sf-http-client/blob/master/test1.php

It's a bit hard to follow what's going on here and what could be the source of this problem. I *believe* you that you're seeing an issue here, but it looks like this could be related to how this example is mixing blocking and non-blocking dependencies (which is usually not recommended). As an alternative, you may also use https://github.com/clue/reactphp-buzz which is known to work just fine and should provide all the APIs you're using here.

> You can also reproduce it by verifying by hand that `select_stream()` does not return a stream ended with no content. (which is quite obvious)

I can not reproduce this, see also https://github.com/reactphp/event-loop/blob/88aaa77ad3ffce3ff22f1361d88c22ec85a50e23/tests/AbstractLoopTest.php#L65 linked above.

> I could understand my changes may be in conflict with another feature, but I break no test.

I'm leaning towards *if it ain't broke, don't fix it*, that's why I'm trying to figure out *what* is broken in the first place :+1:

**Comment by Nek-** [comment]
> As discussed above, this test should call the event listener function (handleData()) twice. This is covered by a number of tests in this project and the underlying event loop, e.g. reactphp/event-loop:tests/AbstractLoopTest.php@88aaa77#L65 linked above.

My point is that: **no**. The diff I suggested with a loop still calls it only once.

In the second example, everything is non-blocking. (it would fail otherwise, because of [this line](https://github.com/reactphp/stream/blob/master/src/ReadableResourceStream.php#L55) (and I was failing before [a fix inside Symfony](https://github.com/symfony/symfony/issues/34943) actually))

In the test you link ([abstract loop test](https://github.com/reactphp/event-loop/blob/88aaa77ad3ffce3ff22f1361d88c22ec85a50e23/tests/AbstractLoopTest.php#L65)) it works like a charm with any stream since they are not decorated with the [ReadableResourceStream](https://github.com/reactphp/stream/blob/master/src/ReadableResourceStream.php) which is precisely what I want to fix.


> I understand this situation can be frustrating

Oh yeah it is 😬 . I just solved a bug, gave examples. And you're just saying me kindly that I say shit with no proof while I continue to get more and more and I'm just doing it on my free time ... IN ORDER TO HELP :')

**Comment by CharlotteDunois** [comment]
Sorry, but I can't reproduce your test case either. `stream_select` successfully reports a close through a readable event.

Script (taken from your added test):
```php
<?php
require 'vendor/autoload.php';

use React\EventLoop\StreamSelectLoop;
use React\Stream\ReadableResourceStream;

$loop = new StreamSelectLoop();
$stream = fopen(__FILE__, 'r+');

$capturedData = null;
$conn = new ReadableResourceStream($stream, $loop);

$conn->on('data', function ($data) use (&$capturedData) {
    $capturedData = $data;
});

$hasEnded = false;
$conn->on('end', function () use (&$hasEnded) {
    $hasEnded = true;
});

$loop->run();
var_dump($capturedData, $hasEnded);
```

Output:
```
string(492) "<?php
require 'vendor/autoload.php'; [...]"
bool(true)
```

I've run the script on WSL Debian.

**Comment by Nek-** [comment]
Ok, I get it. It looks like Symfony HttpClient is by design totally incompatible with reactphp (and any other system based on php streams) since it implements an event loop itself inside their own streams.

This PR can be closed. (oh wait)
